// react-page-scroller/lib/index.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var Events = _interopRequireWildcard(require("./Events"));

var _utils = require("./utils");

var _usePrevValue = _interopRequireDefault(require("./usePrevValue"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

if (!global._babelPolyfill) {
  require("babel-polyfill");
}

var DEFAULT_ANIMATION_TIMER = 1000;
var DEFAULT_ANIMATION = "ease-in-out";
var DEFAULT_CONTAINER_HEIGHT = "100vh";
var DEFAULT_CONTAINER_WIDTH = "100vw";
var DEFAULT_COMPONENT_INDEX = 0;
var DEFAULT_COMPONENTS_TO_RENDER_LENGTH = 0;
var DEFAULT_ANIMATION_TIMER_BUFFER = 200;
var KEY_UP = 38;
var KEY_DOWN = 40;
var MINIMAL_DELTA_Y_DIFFERENCE = 1;
var DISABLED_CLASS_NAME = "rps-scroll--disabled";
var previousTouchMove = null;
var isScrolling = false;
var isMounted = false;
var isBodyScrollEnabled = true;
var isTransitionAfterComponentsToRenderChanged = false;
var containers = [];

var ReactPageScroller = function ReactPageScroller(_ref) {
  var animationTimer = _ref.animationTimer,
      animationTimerBuffer = _ref.animationTimerBuffer,
      blockScrollDown = _ref.blockScrollDown,
      blockScrollUp = _ref.blockScrollUp,
      children = _ref.children,
      containerHeight = _ref.containerHeight,
      containerWidth = _ref.containerWidth,
      customPageNumber = _ref.customPageNumber,
      handleScrollUnavailable = _ref.handleScrollUnavailable,
      onBeforePageScroll = _ref.onBeforePageScroll,
      pageOnChange = _ref.pageOnChange,
      renderAllPagesOnFirstRender = _ref.renderAllPagesOnFirstRender,
      transitionTimingFunction = _ref.transitionTimingFunction;

  var _useState = (0, _react.useState)(DEFAULT_COMPONENT_INDEX),
      componentIndex = _useState[0],
      setComponentIndex = _useState[1];

  var _useState2 = (0, _react.useState)(DEFAULT_COMPONENTS_TO_RENDER_LENGTH),
      componentsToRenderLength = _useState2[0],
      setComponentsToRenderLength = _useState2[1];

  var prevComponentIndex = (0, _usePrevValue["default"])(componentIndex);
  var scrollContainer = (0, _react.useRef)(null);
  var pageContainer = (0, _react.useRef)(null);
  var lastScrolledElement = (0, _react.useRef)(null);
  var scrollPage = (0, _react.useCallback)(function (nextComponentIndex) {
    if (onBeforePageScroll) {
      onBeforePageScroll(nextComponentIndex);
    }

    pageContainer.current.style.transform = "translate3d(0, " + nextComponentIndex * -100 + "%, 0)";
  }, [onBeforePageScroll]);
  var addNextComponent = (0, _react.useCallback)(function (componentsToRenderOnMountLength) {
    var tempComponentsToRenderLength = 0;

    if (!(0, _utils.isNil)(componentsToRenderOnMountLength)) {
      tempComponentsToRenderLength = componentsToRenderOnMountLength;
    }

    tempComponentsToRenderLength = Math.max(tempComponentsToRenderLength, componentsToRenderLength);

    if (tempComponentsToRenderLength <= componentIndex + 1) {
      if (!(0, _utils.isNil)(children[componentIndex + 1])) {
        tempComponentsToRenderLength++;
      }
    }

    setComponentsToRenderLength(tempComponentsToRenderLength);
  }, [children, componentIndex, componentsToRenderLength]);
  var checkRenderOnMount = (0, _react.useCallback)(function () {
    if (renderAllPagesOnFirstRender) {
      setComponentsToRenderLength(_react["default"].Children.count(children));
    } else if (!(0, _utils.isNil)(children[DEFAULT_COMPONENT_INDEX + 1])) {
      addNextComponent(DEFAULT_COMPONENTS_TO_RENDER_LENGTH + 1);
    }
  }, [addNextComponent, children, renderAllPagesOnFirstRender]);
  var disableScroll = (0, _react.useCallback)(function () {
    if (isBodyScrollEnabled) {
      isBodyScrollEnabled = false;
      window.scrollTo({
        left: 0,
        top: 0,
        behavior: "smooth"
      });
      document.body.classList.add(DISABLED_CLASS_NAME);
      document.documentElement.classList.add(DISABLED_CLASS_NAME);
    }
  }, []);
  var enableDocumentScroll = (0, _react.useCallback)(function () {
    if (!isBodyScrollEnabled) {
      isBodyScrollEnabled = true;
      document.body.classList.remove(DISABLED_CLASS_NAME);
      document.documentElement.classList.remove(DISABLED_CLASS_NAME);
    }
  }, []);
  var setRenderComponents = (0, _react.useCallback)(function () {
    var newComponentsToRender = [];
    var i = 0;

    while (i < componentsToRenderLength && !(0, _utils.isNil)(children[i])) {
      containers[i] = true;
      newComponentsToRender.push( /*#__PURE__*/_react["default"].createElement("div", {
        key: i,
        style: {
          height: "100%",
          width: "100%"
        }
      }, _react["default"].cloneElement(children[i], _extends({}, children[i].props))));
      i++;
    }

    return newComponentsToRender;
  }, [children, componentsToRenderLength]);
  var scrollWindowDown = (0, _react.useCallback)(function () {
    if (!isScrolling && !blockScrollDown) {
      if (!(0, _utils.isNil)(containers[componentIndex + 1])) {
        disableScroll();
        isScrolling = true;
        scrollPage(componentIndex + 1);
        setTimeout(function () {
          if (isMounted) {
            setComponentIndex(function (prevState) {
              return prevState + 1;
            });
          }
        }, animationTimer + animationTimerBuffer);
      } else {
        enableDocumentScroll();

        if (handleScrollUnavailable) {
          handleScrollUnavailable();
        }
      }
    }
  }, [animationTimer, animationTimerBuffer, blockScrollDown, componentIndex, disableScroll, enableDocumentScroll, handleScrollUnavailable, scrollPage]);
  var scrollWindowUp = (0, _react.useCallback)(function () {
    if (!isScrolling && !blockScrollUp) {
      if (!(0, _utils.isNil)(containers[componentIndex - 1])) {
        disableScroll();
        isScrolling = true;
        scrollPage(componentIndex - 1);
        setTimeout(function () {
          if (isMounted) {
            setComponentIndex(function (prevState) {
              return prevState - 1;
            });
          }
        }, animationTimer + animationTimerBuffer);
      } else {
        enableDocumentScroll();

        if (handleScrollUnavailable) {
          handleScrollUnavailable();
        }
      }
    }
  }, [animationTimer, animationTimerBuffer, blockScrollUp, componentIndex, disableScroll, enableDocumentScroll, handleScrollUnavailable, scrollPage]);
  var touchMove = (0, _react.useCallback)(function (event) {
    if (!(0, _utils.isNull)(previousTouchMove)) {
      if (event.touches[0].clientY > previousTouchMove) {
        scrollWindowUp();
      } else {
        scrollWindowDown();
      }
    } else {
      previousTouchMove = event.touches[0].clientY;
    }
  }, [scrollWindowDown, scrollWindowUp]);
  var wheelScroll = (0, _react.useCallback)(function (event) {
    if (Math.abs(event.deltaY) > MINIMAL_DELTA_Y_DIFFERENCE) {
      if ((0, _utils.isPositiveNumber)(event.deltaY)) {
        lastScrolledElement.current = event.target;
        scrollWindowDown();
      } else {
        lastScrolledElement.current = event.target;
        scrollWindowUp();
      }
    }
  }, [scrollWindowDown, scrollWindowUp]);
  var keyPress = (0, _react.useCallback)(function (event) {
    if (event.keyCode === KEY_UP) {
      scrollWindowUp();
    }

    if (event.keyCode === KEY_DOWN) {
      scrollWindowDown();
    }
  }, [scrollWindowDown, scrollWindowUp]);
  (0, _react.useEffect)(function () {
    scrollContainer.current.addEventListener(Events.TOUCHMOVE, touchMove);
    scrollContainer.current.addEventListener(Events.KEYDOWN, keyPress);
    return function () {
      scrollContainer.current.removeEventListener(Events.TOUCHMOVE, touchMove);
      scrollContainer.current.removeEventListener(Events.KEYDOWN, keyPress);
    };
  }, [touchMove, keyPress]);
  (0, _react.useEffect)(function () {
    isMounted = true;
    checkRenderOnMount();
    return function () {
      isMounted = false;
    };
  }, []);
  (0, _react.useEffect)(function () {
    isScrolling = false;
    previousTouchMove = null;

    if (componentIndex > prevComponentIndex) {
      addNextComponent();
    }
  }, [addNextComponent, componentIndex, prevComponentIndex]);
  (0, _react.useEffect)(function () {
    if (pageOnChange) {
      pageOnChange(componentIndex);
    }
  }, [pageOnChange, componentIndex]);
  (0, _react.useEffect)(function () {
    if (!(0, _utils.isNil)(customPageNumber) && customPageNumber !== componentIndex) {
      var newComponentsToRenderLength = componentsToRenderLength;

      if (customPageNumber !== componentIndex) {
        if (!(0, _utils.isNil)(containers[customPageNumber]) && !isScrolling) {
          isScrolling = true; // eslint-disable-next-line max-len

          scrollPage(customPageNumber);

          if ((0, _utils.isNil)(containers[customPageNumber]) && !(0, _utils.isNil)(children[customPageNumber])) {
            newComponentsToRenderLength++;
          }

          setTimeout(function () {
            setComponentIndex(customPageNumber);
            setComponentsToRenderLength(newComponentsToRenderLength);
          }, animationTimer + animationTimerBuffer);
        } else if (!isScrolling && !(0, _utils.isNil)(children[customPageNumber])) {
          for (var i = componentsToRenderLength; i <= customPageNumber; i++) {
            newComponentsToRenderLength++;
          }

          if (!(0, _utils.isNil)(children[customPageNumber])) {
            newComponentsToRenderLength++;
          }

          isScrolling = true;
          isTransitionAfterComponentsToRenderChanged = true;
          setComponentsToRenderLength(newComponentsToRenderLength);
        }
      }
    }
  }, [customPageNumber, scrollPage]);
  (0, _react.useEffect)(function () {
    if (isTransitionAfterComponentsToRenderChanged) {
      isTransitionAfterComponentsToRenderChanged = false;
      scrollPage(customPageNumber);
      setTimeout(function () {
        setComponentIndex(customPageNumber);
      }, animationTimer + animationTimerBuffer);
    }
  }, [animationTimer, animationTimerBuffer, componentsToRenderLength, customPageNumber, scrollPage]);
  return /*#__PURE__*/_react["default"].createElement("div", {
    ref: scrollContainer,
    style: {
      height: containerHeight,
      width: containerWidth,
      overflow: "hidden"
    }
  }, /*#__PURE__*/_react["default"].createElement("div", {
    ref: pageContainer,
    onWheel: wheelScroll,
    style: {
      height: "100%",
      width: "100%",
      transition: "transform " + animationTimer + "ms " + transitionTimingFunction,
      outline: "none"
    },
    tabIndex: 0
  }, setRenderComponents()));
};

ReactPageScroller.propTypes = process.env.NODE_ENV !== "production" ? {
  animationTimer: _propTypes["default"].number,
  animationTimerBuffer: _propTypes["default"].number,
  blockScrollDown: _propTypes["default"].bool,
  blockScrollUp: _propTypes["default"].bool,
  children: _propTypes["default"].any,
  containerHeight: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  containerWidth: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  customPageNumber: _propTypes["default"].number,
  handleScrollUnavailable: _propTypes["default"].func,
  onBeforePageScroll: _propTypes["default"].func,
  pageOnChange: _propTypes["default"].func,
  renderAllPagesOnFirstRender: _propTypes["default"].bool,
  transitionTimingFunction: _propTypes["default"].string
} : {};
ReactPageScroller.defaultProps = {
  animationTimer: DEFAULT_ANIMATION_TIMER,
  animationTimerBuffer: DEFAULT_ANIMATION_TIMER_BUFFER,
  transitionTimingFunction: DEFAULT_ANIMATION,
  containerHeight: DEFAULT_CONTAINER_HEIGHT,
  containerWidth: DEFAULT_CONTAINER_WIDTH,
  blockScrollUp: false,
  blockScrollDown: false
};
var _default = ReactPageScroller;
exports["default"] = _default;
module.exports = exports.default;
